File: src/Algorithm/Maze/Movement.java
package src.Algorithm.Maze;

public class Movement {
    private boolean canMove;
    private boolean isDiagonal;
    private String leftOrRightDiagonal;
    private Cell cellToMoveToFirst;
    private int[] direction;

    public Movement(boolean canMove, int[] direction) {
        this.canMove = canMove;
        this.isDiagonal = false;
        this.leftOrRightDiagonal = "";
        this.cellToMoveToFirst = null;
        this.direction = direction;
    }

    public Movement(boolean canMove, boolean isDiagonal, String leftOrRightDiagonal, Cell cellToMoveToFirst, int[] direction) {
        this.canMove = canMove;
        this.isDiagonal = isDiagonal;
        this.leftOrRightDiagonal = leftOrRightDiagonal;
        this.cellToMoveToFirst = cellToMoveToFirst;
        this.direction = direction;
    }

    public boolean getCanMove() {
        return this.canMove;
    }

    public boolean getIsDiagonal() {
        return this.isDiagonal;
    }

    public String getIsLeftRight() {
        return this.leftOrRightDiagonal;
    }

    public Cell getFirstMove() {
        return this.cellToMoveToFirst;
    }

    public int[] getDirection() {
        return this.direction;
    }

    public void setCanMove(boolean canMove) {
        this.canMove = canMove;
    }

    public void setIsDiagonal(boolean isDiagonal) {
        this.isDiagonal = isDiagonal;
    }

    public void setLeftOrRightDiagonal(String leftOrRightDiagonal) {
        this.leftOrRightDiagonal = leftOrRightDiagonal;
    }

    public void setCellToMoveToFirst(Cell cellToMoveToFirst) {
        this.cellToMoveToFirst = cellToMoveToFirst;
    }

    public void setDirection(int[] direction) {
        this.direction = direction;
    }
}

File: src/Algorithm/Maze/Movement.cpp
#include "Movement.h"
#include "Cell.h"
#include <stdexcept>

// Constructor: Movement(bool canMove, const std::array<int, 2>& direction)
Movement::Movement(bool canMove, const std::array<int, 2>& direction)
    : canMove(canMove),
      isDiagonal(false),
      leftOrRightDiagonal(""),
      cellToMoveToFirst(nullptr),
      direction(direction) {}

// Constructor: Movement(bool canMove, bool isDiagonal, const std::string& leftOrRightDiagonal, Cell* cellToMoveToFirst, const std::array<int, 2>& direction)
Movement::Movement(bool canMove, bool isDiagonal, const std::string& leftOrRightDiagonal, Cell* cellToMoveToFirst, const std::array<int, 2>& direction)
    : canMove(canMove),
      isDiagonal(isDiagonal),
      leftOrRightDiagonal(leftOrRightDiagonal),
      cellToMoveToFirst(cellToMoveToFirst),
      direction(direction) {}

// Getter Methods

bool Movement::getCanMove() const {
    return this->canMove;
}

bool Movement::getIsDiagonal() const {
    return this->isDiagonal;
}

std::string Movement::getIsLeftRight() const {
    return this->leftOrRightDiagonal;
}

Cell* Movement::getFirstMove() const {
    return this->cellToMoveToFirst;
}

std::array<int, 2> Movement::getDirection() const {
    return this->direction;
}

// Setter Methods

void Movement::setCanMove(bool canMove) {
    this->canMove = canMove;
}

void Movement::setIsDiagonal(bool isDiagonal) {
    this->isDiagonal = isDiagonal;
}

void Movement::setLeftOrRightDiagonal(const std::string& leftOrRightDiagonal) {
    this->leftOrRightDiagonal = leftOrRightDiagonal;
}

void Movement::setCellToMoveToFirst(Cell* cellToMoveToFirst) {
    this->cellToMoveToFirst = cellToMoveToFirst;
}

void Movement::setDirection(const std::array<int, 2>& direction) {
    this->direction = direction;
}